utils/catalog-parser/src/main/java/de/sovity/edc/utils/catalog/mapper/DspDataOfferBuilder.java

```diff
@@ -65,7 +75,65 @@ private DspDataOffer buildDataOffer(JsonObject dataset) {
     }
 
     @NotNull
-    private DspContractOffer buildContractOffer(JsonObject json) {
-        return new DspContractOffer(JsonLdUtils.id(json), json);
+    DspContractOffer buildContractOffer(JsonObject json) {
+        /*
+         * /!\ Workaround
+         *
+         * The Eclipse EDC uses a new random UUID for each policy that it returns and in turn a new contract ID.
+         * This Eclipse ID can't be used as such.
+         * As a workaround, we must introduce our own ID.
+         * For a first iteration, we will assume that the content of the policy remains the same (same content, same order)
+         * and hash it to use it as a key.
+         */
+
+        // FIXME: This doesn't enforce any property order and may cause trouble if the returned policy schema is not consistent
+        //  Use canonical form if needed later.
+        val noId = Json.createObjectBuilder(json).remove(Prop.ID).build();
+        val policyJsonString = JsonUtils.toJson(noId);
+        val sha1 = sha1(policyJsonString);
+        // encoding with base16 to make the hash readable to humans when decoding with ContractId
+        val humanReadableSha1 = toBase16String(sha1);
+        // re-encoding as base64 because this is the format that ContractId expects
+        val policyId = toBase64String(humanReadableSha1);
```

Comment:
I don't know if you agree, but I'd say the mental complexity of the method is still high, for what we actually do.

What do you think about:

```java
val id = JsonLdUtils.string(json, Prop.ID);
val noId = Json.createObjectBuilder(json).remove(Prop.ID).build();

val hashed = hashJsonObject(noId);
val stableId = stableContractId(id, hashed);

val copy = Json.createObjectBuilder(noId).add(Prop.ID, stableId).build();
return new DspContractOffer(stableId, copy);
```